configure.ac

```diff
@@ -735,6 +736,15 @@ AC_SUBST([gap_kernel_minor_version])
 AC_DEFINE_UNQUOTED([GAP_KERNEL_MAJOR_VERSION], $gap_kernel_major_version, [The major version of the kernel ABI])
 AC_DEFINE_UNQUOTED([GAP_KERNEL_MINOR_VERSION], $gap_kernel_minor_version, [The minor version of the kernel ABI])
 
+dnl libgap_current/revision/age must be set using the GNU Libtool rules at
+dnl https://www.gnu.org/software/libtool/manual/libtool.html#Libtool-versioning
+dnl which uses current:revision:age format.
+dnl Set it up as discussed on GAP's PR 3928
+libgap_current=$(( gap_kernel_major_version + gap_kernel_minor_version ))
```

Comment:
Nope, that won't do it. With that setup,  kernel  3.1 and 4.0 would have the same `current` value.

It really has to be as I described it, not the other way around.

Reply:
It is the way you described it, execept that the libtool numbers are expressed in kernel numbers,
and not the other way around. An invertible linear transformation, how can this get different?

E.g. for kernel 4.0 you'll get Linux libraries libgap.so.4.0.0, and for kernel 3.1 you will get libgap.so.3.1.0. The beauty of libtool versioning :-)

Reply:
That reasoning is flawed, I am afraid, but you are not the first to fall in the trap (I was there, too).

If it was just a matter of identifying kernel versions with library versions, we could use Libtool's `-version-number major[:minor[:revision]]` interface and be done with it! But they caution against it, and there is a reason: it is difficult to map such triples correctly to all different operating systems, which use very different schemes for how to indicate shared library compatibility. The advantage of the current/revision/age approach is that one gets a monotonic increasing "version number" current, which can be compared like that, and be fitted into any system.


We are considering the following twospaces:
1. K := N x N,  containing pairs (kernel major, kernel minor)
2. L := { (c,a) \in N x N |  c >= a } ,  containing pairs (libtool current, libtool age)
and the maps
 f: L -> K,  (c,a) -> (c-a, a)    (mine)
 g: K -> L,  (M,m) -> (M+m, m)     (yours)

Now, our maps must preserve the *semantics* of the versions.  In your scheme, as I said, the kernel versions 3.1 and 4.0 are mapped to libtool current.age pairs  (4,1) and (4,0).  That indicates that the two are (a) binary compatible (same "current") version, and that the first is *newer* than the second. But that's wrong!

The map f of course also would have this issue, except it does *not* if one follows the rules for dealing with libtool current/age pairs: whenever current is incremented, then either age is incremented, too (meaning we stay backwards compatible), or else it is reset to 0 (meaning a breaking change, not backwards compatible).
Then   we might have a progression like this (left c/a pairs, right side their image under f):
  1,0 -> 1.0
  2,1 -> 1.1
  3,2 -> 1.2
  4,0 -> 4.0
  4,1 -> 4.1

And this preserves ordering semantics in the images, too.

Ah well. I guess in the time it took me to write all these explanations, I could have just implemented this myself *sigh*

Reply:
well, either this, or https://github.com/gap-system/gap/pull/3928#issuecomment-604420578, is wrong. Summing up the equations you wrote in the latter one comes to the very expression for libgap_current you object against.

Reply:
I can quite belive I was wrong, sorry.

Reply:
It boils down to the rule one needs to use to increment version. With K, whenever backward compatibility after kernel version (x,y) is to be broken, the next should be (x+y+1,0). E.g. after (x,y)=(3,1) the next kernel should be (5,0), and not (4,0).

I can either add this in comments, or indeed, re-implement the way you prefer, i.e. use `f` rather than `g`.
Please shout your preference.

Reply:
@dimpase no, the comment you quote is correct -- it describes the use of the function `f`. As I explained above, this is NOT equivalent to function `g`, which your PR currently uses.

And yes, of course we can implement semantic versioning for the kernel version on its own. That's what we are already doing. But then there is NO WAY to map this to libtool versions. Your way (using function `g`) is incorrect, as it breaks the ordering. This is not a matter of taste, but one of wrong vs right.

Reply:
I don't understand, I think that's an elementary maths that's a matter of taste, and result 
is the same. (you probably didn't notice the extra rule on kernel version bumping that needs to be imposed to achieve equivalence).

Claim: using g with the rule (D) that (x,y) in K is bumped either to (x,y+1) [compatibility case] or to (x+y+1,0), is equivalent to using  f with the rule (M, aka libtool) that (c,a) in L is bumped to either (c+1,a+1) [compatibility case] or to (c+1,0). 

Indeed, f=[[1,-1],[0,1]] and g=[[1,1],[0,1]] are inverses of each other on RxR, we only need to check that the bumping leads to the same thing. If (c,a) is bumped to (c+1,a+1) [compatibility case] then the kernel version f(c,a)=(c-a,a) is bumped to  f(c+1,a+1)=(c-a,a+1), as'd be by (D).
If (x,y) in K is bumped to (x,y+1) then g(x,y+1)=(x+y+1,y+1), as'd be by (M) - we bump
(x+y,y) to (x+y+1,y+1).
 
If (c,a) is bumped to (c+1,0) then the kernel version f(c,a)=(c-a,a) is bumped to
f(c+1,0)=(c+1,0)=((c-a)+a+1,0), as it should by (D).
If (x,y) in K is bumped to (x+y+1,0) then g(x+y+1,0)=(x+y+1,0), as it'd be by (M) - we bump (x+y,y) to (x+y+1,0).     QED.

Reply:
fixed